src/main/cpp/src/datetime_parser.cu

```diff
@@ -0,0 +1,672 @@
+            segments[i]            = current_segment_value;
+            current_segment_value  = 0;
+            current_segment_digits = 0;
+            i += 1;
+          } else {
+            return ParseResult::INVALID;
+          }
+        } else if (5 == i || 6 == i) {
+          if ('.' == b && 5 == i) {
+            if (!is_valid_digits(i, current_segment_digits)) { return ParseResult::INVALID; }
+            segments[i]            = current_segment_value;
+            current_segment_value  = 0;
+            current_segment_digits = 0;
+            i += 1;
+          } else {
+            if (!is_valid_digits(i, current_segment_digits) || !allow_tz_in_date_str) {
+              return ParseResult::INVALID;
+            }
+            segments[i]            = current_segment_value;
+            current_segment_value  = 0;
+            current_segment_digits = 0;
+            i += 1;
+            *parsed_tz_ptr = bytes + j;
+            // strip the whitespace between timestamp and timezone
+            while (*parsed_tz_ptr < end_ptr && is_whitespace(**parsed_tz_ptr))
+              ++(*parsed_tz_ptr);
+            *parsed_tz_length = end_ptr - *parsed_tz_ptr;
+            break;
+          }
+          if (i == 6 && '.' != b) { i += 1; }
+        } else {
+          if (i < segments_len && (':' == b || ' ' == b)) {
+            if (!is_valid_digits(i, current_segment_digits)) { return ParseResult::INVALID; }
+            segments[i]            = current_segment_value;
+            current_segment_value  = 0;
+            current_segment_digits = 0;
+            i += 1;
+          } else {
+            return ParseResult::INVALID;
+          }
+        }
+      } else {
+        if (6 == i) { digits_milli += 1; }
+        // We will truncate the nanosecond part if there are more than 6 digits,
+        // which results in loss of precision
+        if (6 != i || current_segment_digits < 6) {
+          current_segment_value = current_segment_value * 10 + parsed_value;
+        }
+        current_segment_digits += 1;
+      }
+      j += 1;
+    }
+
+    if (!is_valid_digits(i, current_segment_digits)) { return ParseResult::INVALID; }
+    segments[i] = current_segment_value;
+
+    while (digits_milli < 6) {
+      segments[6] *= 10;
+      digits_milli += 1;
+    }
+
+    segments[0] *= year_sign.value_or(1);
+    // above is ported from Spark.
+
+    // copy segments to equivalent kernel timestamp_components
+    // Note: In order to keep above code is equivalent to Spark implementation,
+    //       did not use `timestamp_components` directly to save values.
+    ts_comp->year         = segments[0];
+    ts_comp->month        = static_cast<int8_t>(segments[1]);
+    ts_comp->day          = static_cast<int8_t>(segments[2]);
+    ts_comp->hour         = static_cast<int8_t>(segments[3]);
+    ts_comp->minute       = static_cast<int8_t>(segments[4]);
+    ts_comp->second       = static_cast<int8_t>(segments[5]);
```

Comment:
I feel like segments should be a struct or use enums for index to make this association more clear.

Reply:
`segments` is from Spark code.
Here copy the segments to `ts_comp` structure.
The parser code is ported from Spark, although has magic numbers, it's equivalent to Spark code.
Let's keep it them?